xds/src/main/java/io/grpc/xds/XdsComms2.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Struct;
+import com.google.protobuf.Value;
+import io.envoyproxy.envoy.api.v2.ClusterLoadAssignment;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.LookasideChannelLb.AdsStreamCallback2;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.CheckForNull;
+
+/**
+ * ADS client implementation.
+ */
+// TODO(zdapeng): This is a temporary and easy refactor of XdsComms, will be replaced by XdsClient.
+// Tests are deferred in XdsClientTest, otherwise it's just a refactor of XdsCommsTest.
+final class XdsComms2 {
+  private final ManagedChannel channel;
+  private final Helper helper;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+
+  @CheckForNull
+  private ScheduledHandle adsRpcRetryTimer;
+
+  // never null
+  private BackoffPolicy adsRpcRetryPolicy;
+  // never null
+  private AdsStream adsStream;
+
+  private final class AdsStream {
+    static final String EDS_TYPE_URL =
+        "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment";
+
+    final AdsStreamCallback2 adsStreamCallback2;
+    final StreamObserver<DiscoveryRequest> xdsRequestWriter;
+    final Stopwatch retryStopwatch = stopwatchSupplier.get().start();
+
+    final StreamObserver<DiscoveryResponse> xdsResponseReader =
+        new StreamObserver<DiscoveryResponse>() {
```

Comment:
The implementation of this anonymous class seems too heavy, which makes nests too deep. We should keep anonymous implementations as simple as possible, ideally with only minimal necessary overrides.